src/encryption.rs

```diff
@@ -0,0 +1,798 @@
+            return unsafe {
+                libc::strdup(
+                    CString::new(format!(
+                        "Encryption key manager encounter non-utf8 file name: {}",
+                        err
+                    ))
+                    .unwrap()
+                    .into_raw(),
+                )
+            };
+        }
+    };
+    match key_manager.link_file(src_fname, dst_fname) {
+        Ok(()) => ptr::null(),
+        Err(err) => unsafe {
+            libc::strdup(
+                CString::new(format!(
+                    "Encryption key manager delete file failure: {}",
+                    err
+                ))
+                .unwrap()
+                .into_raw(),
+            )
+        },
+    }
+}
+
+extern "C" fn encryption_key_manager_rename_file(
+    ctx: *mut c_void,
+    src_fname: *const c_char,
+    dst_fname: *const c_char,
+) -> *const c_char {
+    let key_manager = unsafe { &**(ctx as *mut *const dyn EncryptionKeyManager) };
+    let src_fname = match unsafe { CStr::from_ptr(src_fname).to_str() } {
+        Ok(ret) => ret,
+        Err(err) => {
+            return unsafe {
+                libc::strdup(
+                    CString::new(format!(
+                        "Encryption key manager encounter non-utf8 file name: {}",
+                        err
+                    ))
+                    .unwrap()
+                    .into_raw(),
+                )
+            };
+        }
+    };
+    let dst_fname = match unsafe { CStr::from_ptr(dst_fname).to_str() } {
+        Ok(ret) => ret,
+        Err(err) => {
+            return unsafe {
+                libc::strdup(
+                    CString::new(format!(
+                        "Encryption key manager encounter non-utf8 file name: {}",
+                        err
+                    ))
+                    .unwrap()
+                    .into_raw(),
+                )
+            };
+        }
+    };
+    match key_manager.rename_file(src_fname, dst_fname) {
+        Ok(()) => ptr::null(),
+        Err(err) => unsafe {
+            libc::strdup(
+                CString::new(format!(
+                    "Encryption key manager delete file failure: {}",
+                    err
+                ))
+                .unwrap()
+                .into_raw(),
+            )
+        },
+    }
+}
+
+impl DBEncryptionKeyManager {
+    pub fn new(key_manager: Arc<dyn EncryptionKeyManager>) -> DBEncryptionKeyManager {
+        // The size of the raw pointer is of 16 bytes, and it doesn't fit in a C-style pointer.
+        // Allocate a buffer of 16 bytes to store as ctx in C code.
+        let ptr = Arc::into_raw(key_manager);
+        let instance = unsafe {
+            let ctx = libc::malloc(mem::size_of_val(&ptr));
+            *(ctx as *mut *const dyn EncryptionKeyManager) = ptr;
+            crocksdb_ffi::crocksdb_encryption_key_manager_create(
+                ctx,
+                encryption_key_manager_destructor,
```

Comment:
What about keep a pointer of `ctx` in `DBEncryptionKeyManager` and clean the `ctx during `Drop` as the `ctx` is created in the Rust side?

Reply:
`DBEncryptionKeyManager` is just a wrapper holding a c++ shared_ptr. It is dropped immediately after caller calling `Env::new_key_managed_encrypted_env()`. The cleanup of `ctx` happen in destructor of c++ `KeyManagedEncryptedEnv` when it destruct the shared_ptr.

Reply:
Cleanup of  the `Arc` finally happens in `encryption_key_manager_destructor`.